### PR TITLE
IAM Service Account fixes

### LINKED ISSUE #1974: Possible race condition on creating IAM Service Accounts when eksctl creates the namespace
**What happened?**
I feel there's a race condition when creating IAM Service Account and the namespace is created by eksctl.  I run eksctl create cluster using a config file and we see sporadic errors about the namespace is already created.

**What you expected to happen?**
I expected the following to happen:

1. If the namespace does NOT exits, create the namespace, else continue
2. Create the list of IAM service accounts from config

I feel this error is a moot point, because I don't see the issue if the namespace exist before I run eksctl.  We see this error IF eksctl creates the namespace.

**How to reproduce it?**
Hard because this might be a race condition and this issue is very sporadic and please see logs below.  Sorry didn't capture it with debug set.

**Anything else we need to know?**
Running in a docker container:
```
$ uname -a 
Linux 3a2e2cd38b39 4.19.76-linuxkit #1 SMP Thu Oct 17 19:31:58 UTC 2019 x86_64 Linux
```

**Versions**
Please paste in the output of these commands:
```
$ eksctl version
0.15.0
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.11", GitCommit:"d94a81c724ea8e1ccc9002d89b7fe81d58f89ede", GitTreeState:"clean", BuildDate:"2020-03-12T21:08:59Z", GoVersion:"go1.12.17", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.9-eks-502bfb", GitCommit:"502bfb383169b124d87848f89e17a04b9fc1f6f0", GitTreeState:"clean", BuildDate:"2020-02-07T01:31:02Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```

**Logs**
```
2020/03/23 23:53:44 workdir: /workspace
2020/03/23 23:53:44 fetch the helper scripts [master]...
downloading git::https://git-ext.ecd.axway.com/cloud/eks.git//hack?ref=master
2020/03/23 23:53:46 success!
2020/03/23 23:53:46 fetch the eksctl templates [master]...
downloading git::https://git-ext.ecd.axway.com/cloud/eks.git//templates?ref=master
2020/03/23 23:53:47 success!
[ℹ]  eksctl version 0.15.0
[ℹ]  using region us-east-1
[✔]  using existing VPC (vpc-071f4b1e65a795f34) and subnets (private:[subnet-03f20ce03b399ad82 subnet-02273ba83140801fe subnet-0552215211e0ee505] public:[subnet-063288b6bf5d5b24d subnet-02668bef8e029bf7e subnet-05ef50cefb9705a6c])
[!]  custom VPC/subnets will be used; if resulting cluster doesn't function as expected, make sure to review the configuration of VPC/subnets
[ℹ]  using Kubernetes version 1.14
[ℹ]  creating EKS cluster "bq-sample-project" in "us-east-1" region with
[ℹ]  will create a CloudFormation stack for cluster itself and 0 nodegroup stack(s)
[ℹ]  will create a CloudFormation stack for cluster itself and 0 managed nodegroup stack(s)
[ℹ]  if you encounter any issues, check CloudFormation console or try 'eksctl utils describe-stacks --region=us-east-1 --cluster=bq-sample-project'
[ℹ]  CloudWatch logging will not be enabled for cluster "bq-sample-project" in "us-east-1"
[ℹ]  you can enable it with 'eksctl utils update-cluster-logging --region=us-east-1 --cluster=bq-sample-project'
[ℹ]  Kubernetes API endpoint access will use provided values {publicAccess=true, privateAccess=true} for cluster "bq-sample-project" in "us-east-1"
[ℹ]  2 sequential tasks: { create cluster control plane "bq-sample-project", 3 sequential sub-tasks: { associate IAM OIDC provider, 4 parallel sub-tasks: { 2 sequential sub-tasks: { create IAM role for serviceaccount "add-ons/aws-external-dns", create serviceaccount "add-ons/aws-external-dns" }, 2 sequential sub-tasks: { create IAM role for serviceaccount "add-ons/aws-cluster-autoscaler", create serviceaccount "add-ons/aws-cluster-autoscaler" }, 2 sequential sub-tasks: { create IAM role for serviceaccount "add-ons/cert-manager", create serviceaccount "add-ons/cert-manager" }, 2 sequential sub-tasks: { create IAM role for serviceaccount "add-ons/aws-velero", create serviceaccount "add-ons/aws-velero" } }, update cluster VPC endpoint access configuration } }
[ℹ]  building cluster stack "eksctl-bq-sample-project-cluster"
[ℹ]  deploying stack "eksctl-bq-sample-project-cluster"


[ℹ]  building iamserviceaccount stack "eksctl-bq-sample-project-addon-iamserviceaccount-add-ons-aws-velero"
[ℹ]  building iamserviceaccount stack "eksctl-bq-sample-project-addon-iamserviceaccount-add-ons-aws-cluster-autoscaler"
[ℹ]  building iamserviceaccount stack "eksctl-bq-sample-project-addon-iamserviceaccount-add-ons-cert-manager"
[ℹ]  building iamserviceaccount stack "eksctl-bq-sample-project-addon-iamserviceaccount-add-ons-aws-external-dns"
[ℹ]  deploying stack "eksctl-bq-sample-project-addon-iamserviceaccount-add-ons-aws-external-dns"
[ℹ]  deploying stack "eksctl-bq-sample-project-addon-iamserviceaccount-add-ons-cert-manager"
[ℹ]  deploying stack "eksctl-bq-sample-project-addon-iamserviceaccount-add-ons-aws-velero"
[ℹ]  deploying stack "eksctl-bq-sample-project-addon-iamserviceaccount-add-ons-aws-cluster-autoscaler"
[ℹ]  created namespace "add-ons"
[ℹ]  1 error(s) occurred and cluster hasn't been created properly, you may wish to check CloudFormation console
[ℹ]  to cleanup resources, run 'eksctl delete cluster --region=us-east-1 --name=bq-sample-project'
[✖]  namespaces "add-ons" already exists
Error: failed to create cluster "bq-sample-project"
```

### LINKED ISSUE #1976: timeout on Kubernetes API in service account creation
**What happened?**

EKS cluster creation based on a config file containing service accounts (IRSA) sometimes fails with a timeout on the K8s API.

**What you expected to happen?**

The service account creation should only be attempted once the control plane is available.

**How to reproduce it?**

Create a cluster with a config file specifying service accounts.

```
$ cat irsa-namespace.yaml
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig
metadata:
  name: cluster-1
  region: us-east-1
  version: "1.15"
availabilityZones:
  - us-east-1a
  - us-east-1b
vpc:
  cidr: 192.168.0.0/19
iam:
  withOIDC: true
  serviceAccounts:
  - metadata:
      name: sa1
      namespace: default
    attachPolicyARNs:
    # doesn't matter which policy, one is needed to create the IRSA role
    - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
```

```
$ eksctl create cluster --config-file ./irsa-namespace.yaml -v 4
```

This doesn't happen all the times, but running it 5 times in a row should trigger it at least once.

**Anything else we need to know?**

Reproduced with eksctl 0.15.0 Darwin build, and also with a build from master (62fdcdf4) for Mac OS. Using named profile.

**Versions**

```
$ eksctl version
0.15.0

$ kubectl version
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.4", GitCommit:"8d8aa39598534325ad77120c120a22b3a990b5ea", GitTreeState:"clean", BuildDate:"2020-03-12T23:41:24Z", GoVersion:"go1.14", Compiler:"gc", Platform:"darwin/amd64"}
```

**Logs**

Skipping the lines not related to the service accounts:, the error message is in the second to last line:

```
2020-03-24T17:13:53-07:00 [ℹ]  deploying stack "eksctl-cluster-1-addon-iamserviceaccount-default-sa1"
2020-03-24T17:13:53-07:00 [▶]  start waiting for CloudFormation stack "eksctl-cluster-1-addon-iamserviceaccount-default-sa1"
2020-03-24T17:13:53-07:00 [▶]  waiting for CloudFormation stack "eksctl-cluster-1-addon-iamserviceaccount-default-sa1"
2020-03-24T17:14:13-07:00 [▶]  waiting for CloudFormation stack "eksctl-cluster-1-addon-iamserviceaccount-default-sa1"
2020-03-24T17:14:14-07:00 [▶]  done after 20.244375594s of waiting for CloudFormation stack "eksctl-cluster-1-addon-iamserviceaccount-default-sa1"
2020-03-24T17:14:14-07:00 [▶]  completed task: create IAM role for serviceaccount "default/sa1"
2020-03-24T17:14:14-07:00 [▶]  started task: create serviceaccount "default/sa1"
2020-03-24T17:14:44-07:00 [▶]  failed task: create serviceaccount "default/sa1" (will not run other sequential tasks)
2020-03-24T17:14:44-07:00 [▶]  failed task: 2 sequential sub-tasks: { create IAM role for serviceaccount "default/sa1", create serviceaccount "default/sa1" } (will continue until other parallel tasks are completed)
2020-03-24T17:14:44-07:00 [▶]  failed task: 2 sequential sub-tasks: { create IAM role for serviceaccount "default/sa1", create serviceaccount "default/sa1" } (will not run other sequential tasks)
2020-03-24T17:14:44-07:00 [▶]  failed task: 2 sequential sub-tasks: { associate IAM OIDC provider, 2 sequential sub-tasks: { create IAM role for serviceaccount "default/sa1", create serviceaccount "default/sa1" } } (will not run other sequential tasks)
2020-03-24T17:14:44-07:00 [ℹ]  1 error(s) occurred and cluster hasn't been created properly, you may wish to check CloudFormation console
2020-03-24T17:14:44-07:00 [ℹ]  to cleanup resources, run 'eksctl delete cluster --region=us-east-1 --name=cluster-1'
2020-03-24T17:14:44-07:00 [✖]  checking whether namespace "default" exists: Get https://REDACTED.gr7.us-east-1.eks.amazonaws.com/api/v1/namespaces/default: dial tcp 34.239.66.218:443: i/o timeout
Error: failed to create cluster "cluster-1"
```

From a successful attempt, there seems to be a validation of the control plane availability after the service accounts are created:

```
2020-03-24T15:22:51-07:00 [ℹ]  deploying stack "eksctl-cluster-1-addon-iamserviceaccount-default-sa1"
2020-03-24T15:22:51-07:00 [▶]  start waiting for CloudFormation stack "eksctl-cluster-1-addon-iamserviceaccount-default-sa1"
2020-03-24T15:22:51-07:00 [▶]  waiting for CloudFormation stack "eksctl-cluster-1-addon-iamserviceaccount-default-sa1"
2020-03-24T15:23:09-07:00 [▶]  waiting for CloudFormation stack "eksctl-cluster-1-addon-iamserviceaccount-default-sa1"
2020-03-24T15:23:26-07:00 [▶]  waiting for CloudFormation stack "eksctl-cluster-1-addon-iamserviceaccount-default-sa1"
2020-03-24T15:23:26-07:00 [▶]  done after 35.460932053s of waiting for CloudFormation stack "eksctl-cluster-1-addon-iamserviceaccount-default-sa1"
2020-03-24T15:23:26-07:00 [▶]  completed task: create IAM role for serviceaccount "default/sa1"
2020-03-24T15:23:26-07:00 [▶]  started task: create serviceaccount "default/sa1"
2020-03-24T15:23:56-07:00 [ℹ]  created serviceaccount "default/sa1"
2020-03-24T15:23:56-07:00 [▶]  completed task: create serviceaccount "default/sa1"
2020-03-24T15:23:56-07:00 [▶]  completed task: 2 sequential sub-tasks: { create IAM role for serviceaccount "default/sa1", create serviceaccount "default/sa1" }
2020-03-24T15:23:56-07:00 [▶]  completed task: 2 sequential sub-tasks: { create IAM role for serviceaccount "default/sa1", create serviceaccount "default/sa1" }
2020-03-24T15:23:56-07:00 [▶]  completed task: 2 sequential sub-tasks: { associate IAM OIDC provider, 2 sequential sub-tasks: { create IAM role for serviceaccount "default/sa1", create serviceaccount "default/sa1" } }
2020-03-24T15:23:56-07:00 [✔]  all EKS cluster resources for "cluster-1" have been created
2020-03-24T15:23:56-07:00 [▶]  merging kubeconfig files
2020-03-24T15:23:56-07:00 [▶]  setting current-context to ndegory@cluster-1.us-east-1.eksctl.io
2020-03-24T15:23:56-07:00 [✔]  saved kubeconfig as "/Users/ndegory/.kube/config"
2020-03-24T15:23:56-07:00 [▶]  kubectl: "/usr/local/bin/kubectl"
2020-03-24T15:24:12-07:00 [▶]  kubectl version: v1.17.4
2020-03-24T15:24:12-07:00 [▶]  found authenticator: aws-iam-authenticator
2020-03-24T15:24:27-07:00 [ℹ]  kubectl command should work with "/Users/ndegory/.kube/config", try 'kubectl get nodes'
2020-03-24T15:24:27-07:00 [✔]  EKS cluster "cluster-1" in "us-east-1" region is ready
2020-03-24T15:24:27-07:00 [▶]  cfg.json = \
{
    "kind": "ClusterConfig",
    "apiVersion": "eksctl.io/v1alpha5",
    "metadata": {
        "name": "cluster-1",
        "region": "us-east-1",
        "version": "1.15"
    },
    "iam": {
        "serviceRoleARN": "arn:aws:iam::REDACTED:role/eksctl-cluster-1-cluster-ServiceRole-25ZGEW5DU8VY",
        "withOIDC": true,
        "serviceAccounts": [
            {
                "metadata": {
                    "name": "sa1",
                    "namespace": "default",
                    "creationTimestamp": null,
                    "annotations": {
                        "eks.amazonaws.com/role-arn": "arn:aws:iam::REDACTED:role/eksctl-cluster-1-addon-iamserviceaccount-def-Role1-IG0QT2TXQ5DJ"
                    }
                },
                "attachPolicyARNs": [
                    "arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy"
                ],
                "status": {
                    "roleARN": "arn:aws:iam::REDACTED:role/eksctl-cluster-1-addon-iamserviceaccount-def-Role1-IG0QT2TXQ5DJ"
                }
            }
        ]
    },
    "vpc": {
        "id": "vpc-04dbbb1d1d847e360",
        "cidr": "192.168.0.0/19",
        "securityGroup": "sg-04fa285841c3558a5",
        "subnets": {
            "private": {
                "us-east-1a": {
                    "id": "subnet-06ba00d03976c42cd",
                    "cidr": "192.168.8.0/22"
                },
                "us-east-1b": {
                    "id": "subnet-0fc7f06808dcf86e9",
                    "cidr": "192.168.12.0/22"
                }
            },
            "public": {
                "us-east-1a": {
                    "id": "subnet-0e56e0a266e2af2cc",
                    "cidr": "192.168.0.0/22"
                },
                "us-east-1b": {
                    "id": "subnet-0f9ddb939fb4dcd5b",
                    "cidr": "192.168.4.0/22"
                }
            }
        },
        "sharedNodeSecurityGroup": "sg-04191cbdb18be180e",
        "nat": {
            "gateway": "Disable"
        },
        "clusterEndpoints": {
            "privateAccess": false,
            "publicAccess": true
        }
    },
    "availabilityZones": [
        "us-east-1a",
        "us-east-1b"
    ],
    "status": {
        "endpoint": "https://REDACTED.gr7.us-east-1.eks.amazonaws.com",
        "certificateAuthorityData": "REDACTED",
        "arn": "arn:aws:eks:us-east-1:REDACTED:cluster/cluster-1",
        "stackName": "eksctl-cluster-1-cluster"
    }
}
```

### LINKED ISSUE #1992: Deletion of multiple IAM service accounts stops at the first one
**What happened?**

Deleting multiple IAM service accounts with the --config-file and --only-missing arguments only deletes one. Repeating the commands deletes another one, etc..

**What you expected to happen?**

All candidates to deletion should be deleted on the first run.

**How to reproduce it?**

Create a cluster with IAM service accounts, such as specified in this config file:

```
$ cat cluster.yaml
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig
metadata:
  name: cluster-1
  region: us-east-1
  version: "1.15"
availabilityZones:
  - us-east-1a
  - us-east-1b
vpc:
  cidr: 192.168.0.0/19
iam:
  withOIDC: true
  serviceAccounts:
  - metadata:
      name: sa1
      namespace: kube-system
    attachPolicyARNs:
    - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
  - metadata:
      name: sa2
      namespace: kube-system
    attachPolicyARNs:
    - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
  - metadata:
      name: sa3
      namespace: kube-system
    attachPolicyARNs:
    - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
  - metadata:
      name: sa4
      namespace: kube-system
    attachPolicyARNs:
    - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
  - metadata:
      name: sa5
      namespace: kube-system
    attachPolicyARNs:
    - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
  - metadata:
      name: sa6
      namespace: kube-system
    attachPolicyARNs:
    - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
```

Remove all service accounts from the file, and run the deletion command:

```
$ eksctl delete iamserviceaccount --config-file ./cluster.yaml --only-missing --approve -v4
```

Logs below.

**Versions**

```
$ eksctl version
0.16.0
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.0", GitCommit:"9e991415386e4cf155a24b1da15becaa390438d8", GitTreeState:"clean", BuildDate:"2020-03-26T06:16:15Z", GoVersion:"go1.14", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"15+", GitVersion:"v1.15.10-eks-bac369", GitCommit:"bac3690554985327ae4d13e42169e8b1c2f37226", GitTreeState:"clean", BuildDate:"2020-02-26T01:12:54Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```

**Logs**

```
2020-03-30T17:17:30-07:00 [ℹ]  eksctl version 0.16.0
2020-03-30T17:17:30-07:00 [ℹ]  using region us-east-1
...
2020-03-30T17:17:31-07:00 [ℹ]  comparing 0 iamserviceaccounts defined in the given config ("./cluster.yaml") against remote state
2020-03-30T17:17:34-07:00 [ℹ]  iamserviceaccounts "kube-system/sa6" present in the cluster, but missing from the given config
2020-03-30T17:17:34-07:00 [ℹ]  iamserviceaccounts "kube-system/sa2" present in the cluster, but missing from the given config
2020-03-30T17:17:34-07:00 [ℹ]  iamserviceaccounts "kube-system/sa1" present in the cluster, but missing from the given config
2020-03-30T17:17:34-07:00 [ℹ]  iamserviceaccounts "kube-system/sa4" present in the cluster, but missing from the given config
2020-03-30T17:17:34-07:00 [ℹ]  iamserviceaccounts "kube-system/sa5" present in the cluster, but missing from the given config
2020-03-30T17:17:34-07:00 [ℹ]  iamserviceaccounts "kube-system/sa3" present in the cluster, but missing from the given config
2020-03-30T17:17:34-07:00 [ℹ]  combined include rules: kube-system/sa6
2020-03-30T17:17:34-07:00 [ℹ]  1 iamserviceaccount (kube-system/sa6) was included (based on the include/exclude rules)
2020-03-30T17:17:34-07:00 [ℹ]  5 iamserviceaccounts (kube-system/sa1, kube-system/sa2, kube-system/sa3, kube-system/sa4, kube-system/sa5) were excluded (based on the include/exclude rules)
...
2020-03-30T17:17:34-07:00 [ℹ]  1 task: { 2 sequential sub-tasks: { delete IAM role for serviceaccount "kube-system/sa6" [async], delete serviceaccount "kube-system/sa6" } }
2020-03-30T17:17:34-07:00 [▶]  waiting for 1 parallel tasks to complete
2020-03-30T17:17:34-07:00 [▶]  started task: 2 sequential sub-tasks: { delete IAM role for serviceaccount "kube-system/sa6" [async], delete serviceaccount "kube-system/sa6" }
2020-03-30T17:17:34-07:00 [▶]  started task: delete IAM role for serviceaccount "kube-system/sa6" [async]
2020-03-30T17:17:35-07:00 [ℹ]  will delete stack "eksctl-cluster-1-addon-iamserviceaccount-kube-system-sa6"
2020-03-30T17:17:35-07:00 [▶]  completed task: delete IAM role for serviceaccount "kube-system/sa6" [async]
2020-03-30T17:17:35-07:00 [▶]  started task: delete serviceaccount "kube-system/sa6"
2020-03-30T17:17:35-07:00 [ℹ]  deleted serviceaccount "kube-system/sa6"
2020-03-30T17:17:35-07:00 [▶]  completed task: delete serviceaccount "kube-systemsa6"
2020-03-30T17:17:35-07:00 [▶]  completed task: 2 sequential sub-tasks: { delete IAM role for serviceaccount "kube-system/sa6" [async], delete serviceaccount "kube-system/sa6" }
```

### PR DESCRIPTION
**Pull request by ndegory**
### Description

Fixes #1974, fixes #1976, fixes #1992

### Checklist
- [ ] Added tests that cover your change (if possible) N/A
- [ ] Added/modified documentation as required (such as the `README.md`, or the `site/content` directory) N/A
- [x] Manually tested
- [ ] Added labels for change area (e.g. `area/nodegroup`), target version (e.g. `version/0.12.0`) and kind (e.g. `kind/improvement`)
- [x] Make sure the title of the PR is a good description that can go into the release notes

### CODE REVIEW

**Comment by bquenin** [comment]
Any update?

File: pkg/ctl/create/cluster.go [hunk 1]
```diff
@@ -353,6 +353,20 @@ func doCreateCluster(cmd *cmdutils.Cmd, ng *api.NodeGroup, params *cmdutils.Crea
 			return err
 		}
 
+		// tasks depending on the control plane availability
+		tasks := ctl.NewTasksRequiringControlPlane(cfg)
+
+		logger.Info(tasks.Describe())
+		if errs := tasks.DoAllSync(); len(errs) > 0 {
+			logger.Info("%d error(s) occurred and post actions have failed, you may wish to check CloudFormation console", len(errs))
```
[the remark this answers is not in the document]
Comment (ndegory) [reply pkg/ctl/create/cluster.go:361]: thank you for the feedback @martina-if , that seems indeed simpler like that, I'll update the PR as soon as I can validate the fixes.

**Comment by ndegory** [comment]
@martina-if I've rebased and applied your recommendations.

**Comment by martina-if** [comment]
@ndegory Great! Did you get time to test these new changes?

**Comment by ndegory** [comment]
@martina-if, yes, I've ran the tests again (same cluster config as in the bug report) a few times and all deployments succeed.

**Pull request merged**